common/src/gadgets/te_cond_add.rs

```diff
@@ -4,49 +4,40 @@ use ark_ff::{FftField, Field};
 use ark_poly::univariate::DensePolynomial;
 use ark_poly::{Evaluations, GeneralEvaluationDomain};
 use ark_std::{vec, vec::Vec};
-use core::marker::PhantomData;
 
 use crate::domain::Domain;
 use crate::gadgets::booleanity::BitColumn;
 use crate::gadgets::cond_add::{CondAdd, CondAddValues};
 use crate::gadgets::{ProverGadget, VerifierGadget};
-use crate::{const_evals, AffineColumn, Column, FieldColumn};
-
-// Conditional affine addition:
-// if the bit is set for a point, add the point to the acc and store,
-// otherwise copy the acc value
-pub struct TeCondAdd<F: FftField, P: AffineRepr<BaseField = F>> {
-    pub(super) bitmask: BitColumn<F>,
-    pub(super) points: AffineColumn<F, P>,
-    // The polynomial `X - w^{n-1}` in the Lagrange basis
-    pub(super) not_last: FieldColumn<F>,
-    // Accumulates the (conditional) rolling sum of the points
-    pub acc: AffineColumn<F, P>,
-    pub result: P,
-}
+use crate::{const_evals, AffineColumn, Column};
+
+use super::cond_add::{AffineCondAdd, CondAddGen, CondAddValuesGen};
+
+pub type TECondAdd<C> = CondAddGen<Affine<C>>;
+pub type TECondAddValues<C> = CondAddValuesGen<Affine<C>>;
 
-pub struct TeCondAddValues<F: Field, Curve: TECurveConfig<BaseField = F>> {
-    pub bitmask: F,
-    pub points: (F, F),
-    pub not_last: F,
-    pub acc: (F, F),
-    pub _curve: PhantomData<Curve>,
+impl<C: TECurveConfig> AffineCondAdd for Affine<C>
+where
+    <Self as AffineRepr>::BaseField: FftField,
+{
+    type CondAddT = TECondAdd<C>;
 }
 
-impl<F, Curve> CondAdd<F, Affine<Curve>> for TeCondAdd<F, Affine<Curve>>
+impl<F, C> CondAdd<F, Affine<C>> for TECondAdd<C>
 where
     F: FftField,
-    Curve: TECurveConfig<BaseField = F>,
+    C: TECurveConfig<BaseField = F>,
 {
-    type CondAddValT = TeCondAddValues<F, Curve>;
+    type Values = TECondAddValues<C>;
+
     // Populates the acc column starting from the supplied seed (as 0 doesn't work with the addition formula).
     // As the TE addition formula used is not complete, the seed must be selected in a way that would prevent
     // exceptional cases (doublings or adding the opposite point).
     // The last point of the input column is ignored, as adding it would made the acc column overflow due the initial point.
     fn init(
```

Comment:
Here you could check if the seed is inside the group.

Reply:
In general, SW works ~correctly even when the seed is within the prime order subgroup, provided that the cond add do not (maliciously or not) stumble to the identity.

Does TE **strictly** require the seed to be within the prime order subgroup, or could it also operate with a seed outside the prime order subgroup (with some failure cases that are more or less probable)?
Specifically, if for TE a point outside the prime order group is chosen as a seed, what would be the resulting behavior? I haven't yet analyzed the constraint formula for the TE construction...

Reply:
Yeah, there is a slightly higher risk. Given you have enforced PoP, I think the probablity of failure if you choose S outside the group is like 8-9 times higher than if you choose it inside, for the TE case. For

$$P = (x1, y1)$$, 

$$P+Q$$ fails, if $Q$ 's  is inside one of these sets.
![failing-set](https://github.com/user-attachments/assets/372d5630-1803-4238-8ac9-833f42636de5)

You could argue that 8 times of negligible is still negligble. I'd have put a check to prevent choosing out of the prime subgroup here, as it doesn't have performance hit, but I agree it is somehow a subjective, so your choice.